pkg/retriever/bitswapretriever.go

```diff
@@ -1,29 +1,194 @@
+	bstore := bitswaphelpers.NewMultiblockstore()
+	routing := bitswaphelpers.NewIndexerRouting()
+	bsnet := network.NewFromIpfsHost(host, routing)
+	bitswap := client.New(ctx, bsnet, bstore)
+	bsnet.Start(bitswap)
+	bsrv := blockservice.New(bstore, bitswap)
+	return NewBitswapRetrieverFromDeps(bsrv, routing, bstore)
+}
+
+// NewBitswapRetrieverFromDeps is primarily for testing, constructing behavior from direct dependencies
+func NewBitswapRetrieverFromDeps(bsrv blockservice.BlockService, routing IndexerRouting, bstore MultiBlockstore) *BitswapRetriever {
+	return &BitswapRetriever{
+		bstore:       bstore,
+		routing:      routing,
+		blockService: bsrv,
+	}
+}
+
+// Retrieve initializes a new bitswap session
 func (br *BitswapRetriever) Retrieve(ctx context.Context, request types.RetrievalRequest, events func(types.RetrievalEvent)) types.CandidateRetrieval {
-	return &bitswapRetrieval{br, ctx, request, events}
+	return &bitswapRetrieval{br, blockservice.NewSession(ctx, br.blockService), ctx, request, events}
 }
 
 type bitswapRetrieval struct {
 	*BitswapRetriever
-	ctx     context.Context
-	request types.RetrievalRequest
-	events  func(types.RetrievalEvent)
+	bsGetter blockservice.BlockGetter
+	ctx      context.Context
+	request  types.RetrievalRequest
+	events   func(types.RetrievalEvent)
 }
 
+// RetrieveFromCandidates retrieves via go-bitswap backed with the given candidates, under the auspices of a fetcher.Fetcher
 func (br *bitswapRetrieval) RetrieveFromCandidates(candidates []types.RetrievalCandidate) (*types.RetrievalStats, error) {
-	return nil, errors.New("not implemented")
+	phaseStartTime := time.Now()
+	// this is a hack cause we aren't able to track bitswap fetches per peer for now, so instead we just create a single peer for all events
+	bitswapCandidate := types.NewRetrievalCandidate(peer.ID(""), br.request.Cid, metadata.Bitswap{})
+	br.events(events.Started(br.request.RetrievalID, phaseStartTime, types.RetrievalPhase, bitswapCandidate))
+
+	// setup the linksystem to record bytes & blocks written -- since this isn't automatic w/o go-data-transfer
+	totalWritten := uint64(0)
+	blockCount := uint64(0)
+	cb := func(bytesWritten uint64) {
+		// record first byte received
+		if totalWritten == 0 {
+			br.events(events.FirstByte(br.request.RetrievalID, phaseStartTime, bitswapCandidate))
+		}
+		totalWritten += bytesWritten
+		blockCount++
+	}
+	// setup providers for this retrieval
+	br.BitswapRetriever.routing.AddProviders(br.request.RetrievalID, candidates)
+	// setup providers linksystem for this retrieval
+	br.BitswapRetriever.bstore.AddLinkSystem(br.request.RetrievalID, bitswaphelpers.NewByteCountingLinkSystem(&br.request.LinkSystem, cb))
+
+	// copy the link system
+	wrappedLsys := br.request.LinkSystem
+	// replace the opener with a blockservice wrapper (we still want any known adls + reifiers, hence the copy)
+	wrappedLsys.StorageReadOpener = loaderForSession(br.bsGetter)
```

Comment:
this is both hilarious and unfortunate
a sad side effect is that we lose out on the shiny new streaming block reads that go-car/v2/storage gave us; not that it's likely to matter a whole lot, just sad